File: src-tauri/src/models/proofreading.rs
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

#[derive(Debug, Clone, Serialize, FromRow)]
#[serde(rename_all = "camelCase")]
pub struct ProofreadingRuleDto {
    pub id: String,
    pub project_id: Option<String>,
    pub name: String,
    #[serde(rename = "type")]
    pub rule_type: String,
    pub pattern: Option<String>,
    pub config_json: String,
    pub severity: String,
    pub enabled: i64,
    pub created_at: i64,
    pub updated_at: i64,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ProofreadingIssueDto {
    pub id: String,
    pub rule_id: Option<String>,
    pub document_id: String,
    pub paragraph_id: Option<String>,
    pub severity: String,
    pub message: String,
    pub start_offset: Option<i64>,
    pub end_offset: Option<i64>,
    pub suggestion: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateProofreadingRuleInput {
    pub project_id: Option<String>,
    pub name: String,
    #[serde(rename = "type")]
    pub rule_type: String,
    pub pattern: Option<String>,
    pub config_json: Option<String>,
    pub severity: Option<String>,
    pub enabled: Option<i64>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateProofreadingRuleInput {
    pub rule_id: String,
    pub name: Option<String>,
    #[serde(rename = "type")]
    pub rule_type: Option<String>,
    pub pattern: Option<String>,
    pub config_json: Option<String>,
    pub severity: Option<String>,
    pub enabled: Option<i64>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ListProofreadingRulesInput {
    pub project_id: String,
    pub include_builtin: Option<bool>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RunProofreadingOnDocumentInput {
    pub document_id: String,
    pub enabled_only: Option<bool>,
    pub rule_ids: Option<Vec<String>>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RunProofreadingOnProjectInput {
    pub project_id: String,
    pub enabled_only: Option<bool>,
    pub rule_ids: Option<Vec<String>>,
}

File: app/types/proofreading.ts
export type ProofreadingRuleType =
    | 'duplicate_word'
    | 'continuous_punctuation'
    | 'mixed_punctuation'
    | 'long_sentence'
    | 'long_paragraph'
    | 'sensitive_word'
    | 'name_consistency'
    | 'custom_regex'

export type ProofreadingSeverity = 'info' | 'warning' | 'error'

export interface ProofreadingRule {
    id: string
    projectId?: string | null
    name: string
    type: ProofreadingRuleType | string
    pattern?: string | null
    configJson: string
    severity: ProofreadingSeverity | string
    enabled: number
    createdAt: number
    updatedAt: number
}

export interface ProofreadingIssue {
    id: string
    ruleId?: string | null
    documentId: string
    paragraphId?: string | null
    severity: ProofreadingSeverity | string
    message: string
    startOffset?: number | null
    endOffset?: number | null
    suggestion?: string | null
}

export interface CreateProofreadingRuleInput {
    projectId?: string | null
    name: string
    type: ProofreadingRuleType | string
    pattern?: string | null
    configJson?: string | null
    severity?: ProofreadingSeverity | string | null
    enabled?: number | null
}

export interface UpdateProofreadingRuleInput {
    ruleId: string
    name?: string | null
    type?: ProofreadingRuleType | string | null
    pattern?: string | null
    configJson?: string | null
    severity?: ProofreadingSeverity | string | null
    enabled?: number | null
}

export interface ListProofreadingRulesInput {
    projectId: string
    includeBuiltin?: boolean | null
}

export interface RunProofreadingOnDocumentInput {
    documentId: string
    enabledOnly?: boolean | null
    ruleIds?: string[] | null
}

export interface RunProofreadingOnProjectInput {
    projectId: string
    enabledOnly?: boolean | null
    ruleIds?: string[] | null
}

export interface ProofreadingRuleTypeOption {
    value: ProofreadingRuleType
    label: string
    description: string
}

export const proofreadingRuleTypeOptions: ProofreadingRuleTypeOption[] = [
    {value: 'duplicate_word', label: '重复词', description: '检查“的的”“然后然后”等连续重复。'},
    {value: 'continuous_punctuation', label: '连续标点', description: '检查“！！！”“？？？”等过长标点。'},
    {value: 'mixed_punctuation', label: '标点混用', description: '检查中英文标点连用或风格不统一。'},
    {value: 'long_sentence', label: '超长句', description: '按句子长度阈值提示拆句。'},
    {value: 'long_paragraph', label: '超长段落', description: '按段落长度阈值提示拆段。'},
    {value: 'sensitive_word', label: '敏感词', description: '按词表检查需要替换或确认的词。'},
    {value: 'name_consistency', label: '名称一致性', description: '根据人物设定卡检查疑似错写姓名。'},
    {value: 'custom_regex', label: '自定义正则', description: '使用自定义 pattern 命中问题片段。'}
]

export const proofreadingSeverityOptions: Array<{ value: ProofreadingSeverity, label: string }> = [
    {value: 'info', label: '提示'},
    {value: 'warning', label: '警告'},
    {value: 'error', label: '错误'}
]

export function proofreadingRuleTypeLabel(type: string) {
    return proofreadingRuleTypeOptions.find(option => option.value === type)?.label ?? type
}

export function proofreadingSeverityLabel(severity: string) {
    return proofreadingSeverityOptions.find(option => option.value === severity)?.label ?? severity
}

export function isBuiltinProofreadingRule(rule: ProofreadingRule) {
    return !rule.projectId
}
